src/session/user.rs

```diff
@@ -200,22 +196,22 @@ impl User {
     }
 
     pub fn type_(&self) -> BoxedUserType {
-        self.imp().type_.borrow().clone()
+        self.imp().type_.borrow().as_ref().unwrap().to_owned()
     }
 
-    pub fn set_type(&self, type_: UserType) {
-        if self.type_().0 == type_ {
+    pub fn set_type(&self, type_: BoxedUserType) {
+        if matches!(self.imp().type_.borrow().to_owned(), Some(old_type) if old_type == type_) {
```

Comment:
```suggestion
        if matches!(&*self.imp().type_.borrow(), Some(old_type) if old_type == &type_) {
```

Reply:
Oh, sorry, I didn't see it. As `std::option::Option` does this
```rust
impl<T> PartialEq<Option<T>> for Option<T> where
    T: PartialEq<T>, 
```

We could even make it shorter, I think.
```rust
if self.imp().type_.borrow().as_ref() == Some(&type_) {
    return;
}
```
Would also apply to the other one.

Reply:
Oh, nice find.